base/compiler/tfuncs.jl

```diff
@@ -214,6 +214,7 @@ cglobal_tfunc(@nospecialize(fptr)) = Ptr{Cvoid}
 cglobal_tfunc(@nospecialize(fptr), @nospecialize(t)) = (isType(t) ? Ptr{t.parameters[1]} : Ptr)
 cglobal_tfunc(@nospecialize(fptr), t::Const) = (isa(t.val, Type) ? Ptr{t.val} : Ptr)
 add_tfunc(Core.Intrinsics.cglobal, 1, 2, cglobal_tfunc, 5)
+add_tfunc(Core.Intrinsics.have_fma, 1, 1, @nospecialize(x)->Bool, 1)
```

Comment:
should cost be 0 since we expect this to always constant fold?

Reply:
Would inference be aware of the folding, so the function doesn't get penalized for the side of the branch not taken?

More generally, how are branches handled by the inliner?

Reply:
@aviatesk maybe you can chime in here? The value of `have_fma` is not known at inference time, but will be lowered to an LLVM constant.

Reply:
> Would inference be aware of the folding, so the function doesn't get penalized for the side of the branch not taken?
> 
> More generally, how are branches handled by the inliner?

Branch-aware dead code elimination will only happen when inference can find a dead branch as a unreached region.
So, it requires constant-folding on Julia-level, and otherwise the inlinear won't be able to recognize a dead branch, and thus the function will be penalized for both branches.

@maleadt I wonder if we can write a specific tfunc that `llvm_call` to retrieve the constant value at inference time?

Reply:
That wouldn't work, because for the purpose of multiversioning we want to keep the branch until we've cloned the function. We could have it constant-fold in the non-sysimg case, but I'm not sure it's worth the complexity. I also don't think it needs a special `tfunc`; when I hadn't added the intrinsic to `is_pure_intrinsic_infer` inference already replaced it with a constant branch based on the result of a call to the runtime intrinsic (`jl_have_fma`).